examples/pytorch/rgcn/experimental/entity_classify_dist.py

```diff
@@ -0,0 +1,536 @@
+            compute_acc(global_results[global_test_nid], labels[global_test_nid])
+    else:
+        return -1, -1
+
+class NeighborSampler:
+    """Neighbor sampler
+    Parameters
+    ----------
+    g : DGLHeterograph
+        Full graph
+    target_idx : tensor
+        The target training node IDs in g
+    fanouts : list of int
+        Fanout of each hop starting from the seed nodes. If a fanout is None,
+        sample full neighbors.
+    """
+    def __init__(self, g, fanouts, sample_neighbors):
+        self.g = g
+        self.fanouts = fanouts
+        self.sample_neighbors = sample_neighbors
+
+    def sample_blocks(self, seeds):
+        """Do neighbor sample
+        Parameters
+        ----------
+        seeds :
+            Seed nodes
+        Returns
+        -------
+        tensor
+            Seed nodes, also known as target nodes
+        blocks
+            Sampled subgraphs
+        """
+        blocks = []
+        etypes = []
+        norms = []
+        ntypes = []
+        seeds = th.LongTensor(np.asarray(seeds))
+        cur = seeds
+        for fanout in self.fanouts:
+            frontier = self.sample_neighbors(self.g, cur, fanout, replace=True)
+            etypes = self.g.edata[dgl.ETYPE][frontier.edata[dgl.EID]]
+            norm = self.g.edata['norm'][frontier.edata[dgl.EID]]
+            block = dgl.to_block(frontier, cur)
+            block.srcdata[dgl.NTYPE] = self.g.ndata[dgl.NTYPE][block.srcdata[dgl.NID]]
+            block.edata['etype'] = etypes
+            block.edata['norm'] = norm
+            cur = block.srcdata[dgl.NID]
+            blocks.insert(0, block)
+        return seeds, blocks
+
+def run(args, device, data):
+    g, node_feats, num_of_ntype, num_classes, num_rels, \
+        train_nid, val_nid, test_nid, labels, global_val_nid, global_test_nid = data
+
+    fanouts = [int(fanout) for fanout in args.fanout.split(',')]
+    sampler = NeighborSampler(g, fanouts, dgl.distributed.sample_neighbors)
+    # Create DataLoader for constructing blocks
+    dataloader = DistDataLoader(
+        dataset=train_nid.numpy(),
+        batch_size=args.batch_size,
+        collate_fn=sampler.sample_blocks,
+        shuffle=True,
+        drop_last=False)
+
+    valid_sampler = NeighborSampler(g, [-1] * args.n_layers, dgl.distributed.sample_neighbors)
+    # Create DataLoader for constructing blocks
+    valid_dataloader = DistDataLoader(
+        dataset=val_nid.numpy(),
+        batch_size=args.batch_size,
+        collate_fn=valid_sampler.sample_blocks,
+        shuffle=False,
+        drop_last=False)
+
+    test_sampler = NeighborSampler(g, [-1] * args.n_layers, dgl.distributed.sample_neighbors)
+    # Create DataLoader for constructing blocks
+    test_dataloader = DistDataLoader(
+        dataset=test_nid.numpy(),
+        batch_size=args.batch_size,
+        collate_fn=test_sampler.sample_blocks,
+        shuffle=False,
+        drop_last=False)
```

Comment:
getting all neighbors can be very slow.

Reply:
Testing suppose to use full neighbor sample.

Reply:
it's fine. the accuracy difference should be very small. at least the validation shouldn't use full neighbors.

Reply:
If we want to chose testing result based on validation set, I think validation should also use full neighbors.

Reply:
In the current RGCN implementation we use neighbor sampling for validation because computing the validation accuracy is a reasonably frequent operation and including some randomness in neighbor aggregation for model selection is usually OK.  We do offline inference for testing where we take full neighborhood layer-by-layer however.

Reply:
@classicsong said that there is little difference in speed with full-neighbor offline inference and testing with full neighbor DataLoader.